### PR TITLE
WIP: SilverStripe 4 compatibility

### PR DESCRIPTION
**Pull request by kinglozzer**
We have an upcoming project that’ll make heavy use of caching, so wanted to get this prepared. The caveat there is that this isn’t currently in use on any production sites yet, but I thought I’d submit it early to get eyes on it, and give anyone else looking at this module for SS4 the chance to test/feedback.

I saw there’s already an `ss4-upgrade` branch, but that work doesn’t look complete and hasn’t been touched for a while. I also took a different approach in a couple of places (will add as comments inline).

---

There are two commits here:

1. Cache locking to prevent race conditions on high concurrency websites
2. The actual update for SS4

Ideally the cache locking will be abstracted out so it doesn’t _have_ to use `flock()`, but I gave that a low priority given the number of filesystem ops SilverStripe does on every request anyway.

I’ve not looked at the `cacheinclude-manager` addon yet.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -168,17 +168,36 @@ sites.
 
 ### Enabling
 
-To enable the full request cache the `RequestCache` service needs to be added to the `RequestProcessor` as a filter.
+To enable the full request cache, the `RequestCacheMiddleware` needs to be applied and a `Global` config block needs to be created:
```
Comment (kinglozzer) [inline README.md]: Replaced `RequestCache` (based on SS3’s `RequestFilter` API) with a middleware-based implementation. It’s functionally equivalent

File: README.md [hunk 2]
```diff
@@ -245,12 +252,6 @@ Injector:
         'hello': 'Something'
 ```
 
-### Extra performance
-
-For even more performance from global caching, a new `main.php` file is provided which ensures database connections aren't made when a cache is available for the current request.
```
Comment (kinglozzer) [inline README.md]: This has been removed for now. SilverStripe’s authentication middleware (which populates `Security::getCurrentUser()`) makes a database query anyway, so I don’t think we can save anything there. That aside, the “boot” process is much faster now anyway

File: _config/config.yml [hunk 3]
```diff
@@ -1,106 +1,73 @@
...
         - "request.httpMethod() == 'GET'"
-
+---
+Name: cacheinclude-cache
+---
+SilverStripe\Core\Injector\Injector:
+  Psr\SimpleCache\CacheInterface.CacheInclude:
+    factory: SilverStripe\Core\Cache\CacheFactory
+    constructor:
+      namespace: 'CacheInclude'
```
Comment (kinglozzer) [inline _config/config.yml]: Replaced usage of `doctrine/cache` with Symfony’s cache component included in framework. The main reasons are ease of configuration, and the fact that it automatically provides cache segmentation for draft/live staging modes

File: composer.json [hunk 4]
```diff
@@ -1,31 +1,30 @@
...
-    "type": "silverstripe-module",
+    "type": "silverstripe-vendormodule",
     "require": {
-        "php": ">=5.3.2",
-        "composer/installers": "~1.0",
-        "doctrine/cache": "~1.3",
-        "symfony/yaml": "~2.4",
-        "symfony/expression-language": "~2.4",
-        "psr/log": "~1.0"
+        "silverstripe/framework": "^4",
```
Comment (kinglozzer) [inline composer.json]: Note to self: this needs to be changed to `^4.3`. Earlier versions don’t provide automatic cache segmentation for draft/live

File: src/SilverStripe/Controller.php [hunk 5]
```diff
@@ -1,65 +0,0 @@
...
-namespace Heyday\CacheInclude\SilverStripe;
-
-use Heyday\CacheInclude\CacheInclude;
-use CliController;
-
-/**
- * Class Controller
- * @package Heyday\CacheInclude\SilverStripe
- */
-class Controller extends CliController
```
Comment (kinglozzer) [inline src/SilverStripe/Controller.php]: Replaced with a `BuildTask` implementation that provides the same functionality

File: src/SilverStripe/TemplateParserBlockProvider.php [hunk 6]
```diff
@@ -55,16 +57,18 @@ public static function cacheIncludeTemplate($res)
...
+        $includeName = $res['Arguments'][0]['text'];
+        $configName = isset($res['Arguments'][1]) ? $res['Arguments'][1]['text'] : $includeName;
+        $keyCreator = isset($res['Arguments'][2]) ? $res['Arguments'][2]['text'] : "'CacheIncludeKeyCreator'";
+        $cacheInclude = isset($res['Arguments'][3]) ? $res['Arguments'][3]['text'] : "'CacheInclude'";
 
         return <<<PHP
-\$val .= \Injector::inst()->get($cacheInclude)->process(
-   {$res['Arguments'][0]['text']},
+\$val .= \SilverStripe\Core\Injector\Injector::inst()->get({$cacheInclude})->process(
+   {$configName},
```
Comment (kinglozzer) [inline src/SilverStripe/TemplateParserBlockProvider.php]: I need to update the readme for this, but the change looks more complicated than it is:

```
Old method (will still work)
<% cache_include 'Footer' %>

Causes issues with cache names (backslashes):
<% cache_include 'App\Includes\Footer' %>

Solution: provide cache block name separately:
<% cache_include 'App\Includes\Footer', 'Footer' %>
```

**Comment by stevie-mayhew** [comment]
@kinglozzer great work. If you can get the tests running again and update them to include PHP7 then I'd be happy to merge.

File: README.md [hunk 7]
```diff
@@ -245,12 +258,6 @@ Injector:
         'hello': 'Something'
```
Comment (stevie-mayhew) [inline README.md]: Just above at line 255: This needs to be changed from `RequestCache` to `RequestCacheMiddleware`

Do you know how you can comment on lines not within the change set?

Comment (kinglozzer) [reply README.md]: > Do you know how you can comment on lines not within the change set?

No idea 😅

**Comment by kinglozzer** [comment]
> Apart from that I've got it up and running on a site and everything seems stable so I'd be happy to merge and tag

Sweet! Feel free to hold off on tagging a release if you want to play safe - I know I’ll be doing a lot more testing over the coming weeks so edge cases may crop up 😅

**Comment by stevie-mayhew** [comment]
@kinglozzer thanks for this, really great.

I'll hold off tagging and splitting branches for a couple of weeks while we both use it :)

**Comment by stevie-mayhew** [comment]
@kinglozzer anything to report back for usage of this over the past few weeks?

We're running this in production on a large web instance now and are seeing good results so I am pretty ready to tag v5.0.0 if you don't have any objections?

**Comment by kinglozzer** [comment]
I’m afraid not, as it’s still not live! It’s working in testing though, so no objections to tagging here 😅
